### PR TITLE
Move menhir rules to own module

### PR DESCRIPTION
**Pull request by rgrinberg**
I haven't changed the rules at all yet. So things are still a bit ugly. But at least we can keep developing things independently. It would be nice if `Menhir.to_rules` would call `user_rules`, but this would introduce a circular dependency with `Gen_rules`.

### CODE REVIEW

**Comment by diml** [comment]
Actually, the next step is to replace `to_rules` by a more classic `gen_rules : Super_context.t -> unit` pattern and use the `Build` directly rather than generate a `rule` stanza.

**Comment by rgrinberg** [comment]
@diml I pushed a commit to implement the menhir rules without relying on user_rules.

File: src/menhir.ml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+open Import
+
+let gen_rules sctx ~dir ~scope (t : Jbuild.Menhir.t) =
+  let targets n = List.map ~f:(Path.relative dir) [n ^ ".ml"; n ^ ".mli"] in
+  let flags =
+    List.map ~f:(Super_context.expand_vars sctx ~scope ~dir) t.flags in
+  let menhir ~extra_targets =
+    Build.run ~extra_targets
+      (Super_context.resolve_program sctx ~hint:"opam install menhir" "menhir")
```
Comment (diml) [inline src/menhir.ml]: This should be outside of the `let menhir` function, so that it's cached between the calls

File: src/menhir.ml [hunk 2]
```diff
@@ -0,0 +1,32 @@
+open Import
+
+let gen_rules sctx ~dir ~scope (t : Jbuild.Menhir.t) =
+  let targets n = List.map ~f:(Path.relative dir) [n ^ ".ml"; n ^ ".mli"] in
+  let flags =
+    List.map ~f:(Super_context.expand_vars sctx ~scope ~dir) t.flags in
```
Comment (diml) [inline src/menhir.ml]: For uniformity with the rest of Jbuilder, we should change that so that flags use an unexpanded ordered set language

Comment (rgrinberg) [reply src/menhir.ml]: OK, I was meaning to do that anyway. Just didn't want to introduce new behavior in a refactor.

**Comment by rgrinberg** [comment]
@diml improved the flags field and made resolve_program evaluate only once.

**Comment by diml** [comment]
Ok, I added some doc to the code, and to `arg_spec.mli` as well while I was at it

**Pull request merged**